### PR TITLE
turn QuantumScript.hash into a cached property

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

When investigating some performance issues, I noticed that we weren't caching `QuantumScript.hash`, even though calculating it is a known performance bottleneck.

**Description of the Change:**

Make `QuantumScript.hash` a cached property.

**Benefits:**

Much lower classical overhead when we are using caching.

Screenshot 2024-06-28 at 5 38 15 PM

**Possible Drawbacks:**

If someone does bad things and mutates the tape in place, they will have to deal with the resulting consequences.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/tape/qscript.py [hunk 1]
```diff
@@ -188,7 +188,7 @@ def __init__(
     def __repr__(self):
         return f"<{self.__class__.__name__}: wires={self.wires.tolist()}, params={self.num_params}>"
 
-    @property
+    @cached_property
```
Comment (dwierichs) [inline pennylane/tape/qscript.py:191]: Seeing this makes me wonder whether we should be using `cached_property` in more places (throughout PL), where we manually create an attribute `_...` and compute it only upon the first call to the corresponding method/property. :thinking:

File: pennylane/tape/qscript.py [hunk 2]
```diff
@@ -383,6 +383,7 @@ def _update(self):
             # Invalidate cached properties so they get recalculated
             del self.wires
             del self.par_info
+            del self.hash
```
Comment (dwierichs) [inline pennylane/tape/qscript.py:386]: Would be good to update the `test_cached_properties` test to include the hash.
Do we need additional tests to verify that caching the hash works and does not equate unequal tapes in some way?

**Review by dwierichs — CHANGES_REQUESTED** [review]
Nice and simple upgrade @albi3ro ! :tada: 

I think it might be good to update and/or extend the tests to check the caching and to test that the hash is not maintained between unequal tapes created via `_update_...` calls.

File: pennylane/tape/qscript.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/tape/qscript.py:191]: Definitely!

It does rely on `__eq__` and `__hash__`, so if calculating those can get expensive (ie operator), it's probably worth still doing it manually.

File: pennylane/tape/qscript.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/tape/qscript.py:387]: thanks for the reminder. turns out we weren't refreshing `trainable_params` :(

File: pennylane/tape/qscript.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/tape/qscript.py:191]: Perfect, thanks!

File: pennylane/tape/qscript.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/tape/qscript.py:387]: Thanks for fixing it!

File: tests/tape/test_qscript.py [hunk 3]
```diff
@@ -199,11 +199,18 @@ def test_cached_properties(self):
         ]
         m = [qml.expval(qml.Hermitian(2 * np.eye(2), wires=0))]
         qs = QuantumScript(ops, m)
+        assert qs.wires == qml.wires.Wires((0,))
+        qs.par_info  # trigger caching # pylint: disable=pointless-statement
```
Comment (dwierichs) [inline tests/tape/test_qscript.py:203]: Could optionally do
```suggestion
        assert isinstance(qs.par_info, list) and len(qs.par_info) > 0
```

File: tests/tape/test_qscript.py [hunk 4]
```diff
@@ -199,11 +199,18 @@ def test_cached_properties(self):
...
+        qs.hash  # trigger caching # pylint: disable=pointless-statement
+        assert qs.trainable_params == [0, 1, 2, 3, 4, 5, 6, 7]
+
         qs._ops = []
         qs._measurements = []
         qs._update()
         assert qs.wires == qml.wires.Wires([])
         assert isinstance(qs.par_info, list) and len(qs.par_info) == 0
```
Comment (dwierichs) [inline tests/tape/test_qscript.py:211]: Smells like
```suggestion
        assert qs.par_info == []
```

File: tests/tape/test_qscript.py [hunk 5]
```diff
@@ -199,11 +199,18 @@ def test_cached_properties(self):
...
+        assert QuantumScript([], []).hash == qs.hash
```
Comment (dwierichs) [inline tests/tape/test_qscript.py:212]:
```suggestion
        old_hash = qs.hash
        assert qs.trainable_params == [0, 1, 2, 3, 4, 5, 6, 7]

        qs._ops = []
        qs._measurements = []
        qs._update()
        assert qs.wires == qml.wires.Wires([])
        assert isinstance(qs.par_info, list) and len(qs.par_info) == 0
        assert QuantumScript([], []).hash == qs.hash and qs.hash != old_hash
```

**Review by dwierichs — APPROVED** [review]
Thanks for updating the test.
Had some minor suggestions, but feel free to discard them.

**Review by lillian542 — APPROVED** [review]

**Pull request merged**